File: backend-orchestrator/internal/mcp/connector_deploying_test.go
package mcp

import (
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"
)

// KI-FIRST-CONNECTION-TEST-FALLS-BACK-TO-AN-UNUSABLE-STDIO-INTERPRETER: the first Test
// Connection for a never-deployed connector fell through to a stdio subprocess on the
// orchestrator's own interpreter and showed the user "No module named 'pymongo'".

func TestFriendlyTestConnectionErrorMapsImportErrorsToStillSettingUp(t *testing.T) {
	want := ConnectorDeployingMessage("MongoDB")
	// Pinned verbatim: api-gateway connections_deploying_test.go and the frontend's
	// connectorDeploying.test.ts assert on this exact text.
	if want != "The MongoDB connector is still being set up for first use — this can take a minute or two. Please try again shortly." {
		t.Fatalf("unexpected deploying message: %q", want)
	}

	mapped := []string{
		// The raw connector error.
		"No module named 'pymongo'",
		// The #1006-annotated stdio fallback error as observed on 2026-09-16.
		"No module named 'pymongo' [" + stdioFallbackMarker + ": connector mongodb@v1.0.0 ran as a subprocess inside the orchestrator because no Docker container was reachable, and its dependencies could not be installed there (failed to create venv: exit status 1) — this error describes the fallback interpreter, not the connector.]",
		"Connection test failed: ModuleNotFoundError: No module named 'pymongo.errors'",
		// StartServer's own retryable error, already flattened to a string.
		(&ConnectorDeployingError{Connector: "mongodb", Version: "v1.0.0", DisplayName: "MongoDB"}).Error(),
	}
	for _, raw := range mapped {
		got, ok := FriendlyTestConnectionError("MongoDB", raw)
		if !ok || got != want {
			t.Errorf("FriendlyTestConnectionError(%q) = (%q, %v), want (%q, true)", raw, got, ok, want)
		}
		if strings.Contains(got, "No module named") {
			t.Errorf("mapped message still leaks the import error: %q", got)
		}
	}

	// Control: genuine connector failures must pass through untouched — otherwise a bad
	// password would be reported as "still setting up" forever.
	for _, raw := range []string{
		"Authentication failed.",
		"connection refused",
		"Failed to start server: failed to locate connector \"nope\" in tools dir /x",
		"",
	} {
		got, ok := FriendlyTestConnectionError("MongoDB", raw)
		if ok || got != raw {
			t.Errorf("FriendlyTestConnectionError(%q) = (%q, %v), want unchanged", raw, got, ok)
		}
	}

	if got := ConnectorDeployingMessage(""); !IsConnectorDeployingMessage(got) {
		t.Errorf("empty display name lost the marker: %q", got)
	}
}

// writeFakeConnector lays out <tools>/public/database/<id>/{latest.json,versions/v1.0.0/metadata.json}
// with NO connector.py, so reaching the stdio path fails deterministically with
// "connector script not found" instead of spawning a process.
func writeFakeConnector(t *testing.T, id, displayName string) string {
	t.Helper()
	tools := t.TempDir()
	root := filepath.Join(tools, "public", "database", id)
	vdir := filepath.Join(root, "versions", "v1.0.0")
	if err := os.MkdirAll(vdir, 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(root, "latest.json"),
		[]byte(`{"current_version":"v1.0.0","all_versions":["v1.0.0"]}`), 0o644); err != nil {
		t.Fatal(err)
	}
	meta := `{"id":"` + id + `","name":"` + id + `","display_name":"` + displayName + `","version":"1.0.0","connector_type":"` + id + `"}`
	if err := os.WriteFile(filepath.Join(vdir, "metadata.json"), []byte(meta), 0o644); err != nil {
		t.Fatal(err)
	}
	return tools
}

// slowToolGenerator mimics tool-generator in deployer mode: /v1/deploy does not answer
// until a cold image build finishes, which is longer than the orchestrator waits.
func slowToolGenerator(t *testing.T) *httptest.Server {
	t.Helper()
	release := make(chan struct{})
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		select {
		case <-release:
		case <-r.Context().Done():
		}
	}))
	t.Cleanup(func() { close(release); srv.Close() })
	return srv
}

func TestStartServerReturnsStillDeployingInsteadOfStdioWhenDeployCallTimesOut(t *testing.T) {
	const id = "zzfakedeployconn"
	tools := writeFakeConnector(t, id, "Fake DB")
	srv := slowToolGenerator(t)
	t.Setenv("TOOL_GENERATOR_URL", srv.URL)

	oldTimeout := deployCallTimeout
	deployCallTimeout = 200 * time.Millisecond
	t.Cleanup(func() { deployCallTimeout = oldTimeout })

	sm := NewServerManager(tools)
	_, err := sm.StartServer(ServerConfig{
		Name:                  id,
		Version:               "latest",
		NoStdioWhileDeploying: true,
		DeployWaitTimeout:     50 * time.Millisecond,
	})
	if err == nil {
		t.Fatal("expected a still-deploying error, got nil")
	}
	if !IsConnectorDeploying(err) {
		t.Fatalf("expected *ConnectorDeployingError (no stdio fallback), got %T: %v", err, err)
	}
	if want := ConnectorDeployingMessage("Fake DB"); err.Error() != want {
		t.Errorf("message = %q, want %q (display_name from metadata)", err.Error(), want)
	}
	sm.mu.RLock()
	defer sm.mu.RUnlock()
	if s := sm.servers[makeServerKey(id, "v1.0.0")]; s != nil {
		t.Errorf("a server was registered (%s) — the stdio fallback ran", s.ConnType)
	}
}

// Control for the test above: the same timed-out deploy WITHOUT the opt-in keeps the
// existing behavior (poll, then the stdio path — here "connector script not found"),
// and with the opt-in but no tool-generator at all (Docker-less Helm batch) stdio is
// still the transport. Without this, a StartServer that always returned the deploying
// error would pass the test above.
func TestStartServerStillUsesStdioWithoutOptInOrWithoutDeployer(t *testing.T) {
	const id = "zzfakedeployconn"
	tools := writeFakeConnector(t, id, "Fake DB")

	oldTimeout := deployCallTimeout
	deployCallTimeout = 200 * time.Millisecond
	t.Cleanup(func() { deployCallTimeout = oldTimeout })

	t.Run("deploy timed out, no opt-in", func(t *testing.T) {
		srv := slowToolGenerator(t)
		t.Setenv("TOOL_GENERATOR_URL", srv.URL)
		_, err := NewServerManager(tools).StartServer(ServerConfig{
			Name: id, Version: "latest", DeployWaitTimeout: 50 * time.Millisecond,
		})
		if err == nil || IsConnectorDeploying(err) || !strings.Contains(err.Error(), "connector script not found") {
			t.Fatalf("expected the stdio path (connector script not found), got %v", err)
		}
	})

	t.Run("opt-in, no tool-generator configured", func(t *testing.T) {
		t.Setenv("TOOL_GENERATOR_URL", "")
		_, err := NewServerManager(tools).StartServer(ServerConfig{
			Name: id, Version: "latest", NoStdioWhileDeploying: true, DeployWaitTimeout: 50 * time.Millisecond,
		})
		if err == nil || IsConnectorDeploying(err) || !strings.Contains(err.Error(), "connector script not found") {
			t.Fatalf("expected the stdio path (connector script not found), got %v", err)
		}
	})
}

func TestTryDeployConnectorContainerTreatsTimeoutAsInProgress(t *testing.T) {
	oldTimeout := deployCallTimeout
	deployCallTimeout = 100 * time.Millisecond
	t.Cleanup(func() { deployCallTimeout = oldTimeout })

	sm := NewServerManager(t.TempDir())

	srv := slowToolGenerator(t)
	t.Setenv("TOOL_GENERATOR_URL", srv.URL)
	if deployed, building := sm.tryDeployConnectorContainer("mongodb", "v1.0.0"); !deployed || building {
		t.Errorf("timed-out deploy call: got (deployed=%v, building=%v), want (true, false)", deployed, building)
	}

	// Control: a refused connection is a real failure, not a deploy in progress.
	closed := httptest.NewServer(http.NotFoundHandler())
	url := closed.URL
	closed.Close()
	t.Setenv("TOOL_GENERATOR_URL", url)
	if deployed, _ := sm.tryDeployConnectorContainer("mongodb", "v1.0.0"); deployed {
		t.Error("refused deploy call was treated as deployed")
	}
}

File: frontend/src/lib/errors/__tests__/connector-deploying.test.ts
import { describe, it, expect } from "vitest"
import {
  CONNECTOR_DEPLOYING_MARKER,
  connectorDeployingErrorMessage,
  connectorDeployingMessage,
  connectorDeployingSaveNotice,
  isConnectorDeployingResponse,
} from "../connector-deploying"

// Verbatim backend text (backend-orchestrator mcp.ConnectorDeployingMessage("MongoDB")).
const BACKEND_MSG =
  "The MongoDB connector is still being set up for first use — this can take a minute or two. Please try again shortly."

describe("connector-deploying", () => {
  it("builds the same message as the backend", () => {
    expect(connectorDeployingMessage("MongoDB")).toBe(BACKEND_MSG)
    expect(connectorDeployingMessage("")).toContain(CONNECTOR_DEPLOYING_MARKER)
  })

  it("recognizes the gateway's retryable response", () => {
    expect(
      isConnectorDeployingResponse({ success: false, status: "connector_deploying", retryable: true, error: BACKEND_MSG }),
    ).toBe(true)
    // Marker alone (e.g. a gateway that did not add status/retryable).
    expect(isConnectorDeployingResponse({ success: false, status: "failed", error: BACKEND_MSG })).toBe(true)
    // A raw import error from an older backend is the same condition.
    expect(isConnectorDeployingResponse({ success: false, error: "No module named 'pymongo'" })).toBe(true)
  })

  it("does not classify real failures or successes as deploying", () => {
    expect(isConnectorDeployingResponse({ success: false, status: "failed", error: "Authentication failed." })).toBe(false)
    expect(isConnectorDeployingResponse({ success: true, status: "success", message: "Connection test successful" })).toBe(false)
    expect(isConnectorDeployingResponse(null)).toBe(false)
  })

  it("maps raw import errors to the friendly message and leaves other errors alone", () => {
    const raw =
      "No module named 'pymongo' [mcp stdio fallback: connector mongodb@v1.0.0 ran as a subprocess inside the orchestrator because no Docker container was reachable]"
    const mapped = connectorDeployingErrorMessage(raw, "MongoDB")
    expect(mapped).toBe(BACKEND_MSG)
    expect(mapped).not.toContain("No module named")

    expect(connectorDeployingErrorMessage(BACKEND_MSG, "Other")).toBe(BACKEND_MSG)
    expect(connectorDeployingErrorMessage("Connection refused", "MongoDB")).toBeNull()
    expect(connectorDeployingErrorMessage("", "MongoDB")).toBeNull()
  })

  it("maps the gateway's retryable save refusal (edit page Save) to a wait-and-retry notice", () => {
    const body = {
      error: "connector_deploying",
      status: "connector_deploying",
      retryable: true,
      message: "The connector is still being set up, so nothing was saved. Try again in a minute.",
      test_error: BACKEND_MSG,
    }
    expect(connectorDeployingSaveNotice(503, body)).toEqual({
      title: "Connector is still being set up — changes not saved",
      description: BACKEND_MSG,
    })
    // No test_error: fall back to the generic deploying message.
    expect(connectorDeployingSaveNotice(503, { error: "connector_deploying" })?.description).toContain(
      CONNECTOR_DEPLOYING_MARKER,
    )
  })

  it("leaves other save refusals to their normal handling", () => {
    expect(
      connectorDeployingSaveNotice(422, { error: "connection_test_failed", test_error: "Authentication failed." }),
    ).toBeNull()
    expect(connectorDeployingSaveNotice(503, { error: "connection_test_interrupted" })).toBeNull()
    expect(connectorDeployingSaveNotice(503, null)).toBeNull()
  })
})
